WordPress/gutenberg
# Fixed-schedule AutosaveMonitor

## Description

The current AutosaveMonitor does not consistently execute autosaves in response to user input.

This PR proposes a completely different approach. Instead of trying to get scheduling and cancelling right, we:

* Store one boolean flag: `needsAutosave` and flip it to true or false anytime something relevant happens.
* Always run a handler every 60 seconds. If the flag is true, autosave.
* If the handler is running at a time when `isAutosaveable` is false, keep retrying every 1 second.

As a result, autosave will be executed at most 60 seconds after the first unsaved input, sometimes less.

## How has this been tested?

Unit tests are failing now - I'd like to have some discussion about the approach before spending time on adjusting them.

For manual testing, see the instructions below:

Default autosave interval is 60 seconds, if you want to save some time you could modify line 81 to say `interval: 4,` instead of just `interval,`. Then:

* Create a new post or open an existing one.
* Start typing and then stop, confirm the autosave kicks in within a minute (or 4 seconds).
* Confirm there is no subsequent autosave after another minute if you don't type anything again.
* Type something else, confirm the autosave worked.
* Keep typing while the autosave is going and confirm another one happened.
* Type something, save the post manually, don't type anything else, confirm no autosave was triggered.

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
